src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,587 @@
+    }
+}
+
+/// Errors that can occur while writing a log.
+#[derive(Debug)]
+pub enum IntervalLogWriterError<E> {
+    /// Histogram serialization failed.
+    SerializeError(E),
+    /// An i/o error occurred.
+    IoError(io::ErrorKind),
+}
+
+impl<E> From<io::Error> for IntervalLogWriterError<E> {
+    fn from(e: io::Error) -> Self {
+        IntervalLogWriterError::IoError(e.kind())
+    }
+}
+
+/// Write interval logs.
+struct InternalLogWriter<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> {
+    writer: &'a mut W,
+    serializer: &'b mut S,
+    text_buf: String,
+    serialize_buf: Vec<u8>,
+}
+
+impl<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> InternalLogWriter<'a, 'b, W, S> {
+    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
+        self.writer.write_fmt(args)
+    }
+
+    fn write_comment(&mut self, s: &str) -> io::Result<()> {
+        write!(self.writer, "#{}\n", s)
+    }
+
+    fn write_histogram<T: Counter>(
+        &mut self,
+        h: &Histogram<T>,
+        start_timestamp: f64,
+        duration: f64,
+        tag: Option<Tag>,
+        max_value_divisor: f64,
+    ) -> Result<(), IntervalLogWriterError<S::SerializeError>> {
+        self.serialize_buf.clear();
+        self.text_buf.clear();
+
+        if let Some(Tag(s)) = tag {
+            write!(self.text_buf, "Tag={},", &s).expect("Writes to a String can't fail");
+        }
+
+        write!(
+            self.writer,
+            "{}{:.3},{:.3},{:.3},",
+            self.text_buf,
+            start_timestamp,
+            duration,
+            h.max() as f64 / max_value_divisor // because the Java impl does it this way
+        )?;
+
+        self.text_buf.clear();
+        let _len = self.serializer
+            .serialize(h, &mut self.serialize_buf)
+            .map_err(|e| IntervalLogWriterError::SerializeError(e))?;
+        base64::encode_config_buf(&self.serialize_buf, base64::STANDARD, &mut self.text_buf);
+
+        self.writer.write_all(self.text_buf.as_bytes())?;
+        self.writer.write_all(b"\n")?;
+
+        Ok(())
+    }
+}
+
+/// A tag for an interval histogram.
+///
+/// Tags are just `str`s that do not contain a few disallowed characters: ',', '\r', '\n', and ' '.
+///
+/// To get the wrapped `str` back out, use `as_str()` or the `Deref<str>` implementation
+/// (`&some_tag`).
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct Tag<'a>(pub &'a str);
+
+impl<'a> Tag<'a> {
+    /// Create a new Tag.
+    ///
+    /// If a disallowed character is present, this will return `None`.
+    pub fn new(s: &'a str) -> Option<Tag<'a>> {
+        if s.chars()
+            .any(|c| c == ',' || c == '\r' || c == '\n' || c == ' ')
+        {
+            None
+        } else {
+            Some(Tag(s))
+        }
+    }
+
+    /// Returns the tag contents as a str.
+    pub fn as_str(&self) -> &'a str {
+        self.0
+    }
+}
+
+impl<'a> ops::Deref for Tag<'a> {
```

Comment:
You might want to `impl` `AsRef` instead here. Though not clear that it matters.

Reply:
Yeah, this is really poorly documented. Reading a bit more though, I think implementing `Deref` might actually be the right thing after all, so leave it as-is.